### ISSUE TITLE
StreetViewPanorama Directive Development

### ISSUE THREAD

**Issue by rasekar2000** [allenhwkim/angularjs-google-maps#138]
Hi allen,

In Continuation to my issue #133 , I am trying to get all the changes in StreetView that are achieved in jQuery (http://plnkr.co/edit/7S7dIs?p=preview)  in Angular way. I think map directives can passthrough the on- Google API events... (http://plnkr.co/edit/HW9Mkr?p=preview) But its not working.. I guess, I am doing something wrong.. Can you please guide me.

Thanks and Regards,
Raj

**Comment by rasekar2000** [comment]
I understood that these are not Map's  Methods... 
 https://developers.google.com/maps/documentation/javascript/reference#Map
But  StreetViewPanorma's Methods.
https://developers.google.com/maps/documentation/javascript/reference#StreetViewPanorama


So, should we need to create a separate directive for StreetView & hence a way to access its Methods.
	
      <map street-view="StreetViewPanorama(document.querySelector('#street-view'), {position:new google.maps.LatLng(12.925469, 80.250943)})" >

            <streetview  on-position_changed="showSVDetails(this)" on-pov_changed="showSVDetails(this)" on-links_changed="showSVDetails(this)" > 

     </map>

Can you please guide me? Thanks in advance.

Regards,
Raj

**Comment by allenhwkim** [comment]
Yes, you are right.
Unless you do raw coding, we need to have streetview directive.
It will take time for me to implement this since I have limited time to spend for this project.
If you are interested in development for this, please do so.

**Comment by rasekar2000** [comment]
Thanks for the input. Appreciate your openness in time constraint.
If you could give some outline to consider the details, myself or my friends will make an attempt.Then you can merge it to the main branch after your review.
Can you please connect with me to my email - raj.raj.california@gmail.com

**Comment by allenhwkim** [comment]
This is open source code. I will list the requirements here

directive name: `street-view-panorama`
-----------------------------------------------------
requires: `map` directive

* It must set the all [properties](https://developers.google.com/maps/documentation/javascript/reference#StreetViewPanorama) attributes using;

  * Attr2Options#getOptions
    * addressControl
    * clickToGo
    * disableDefaultUI
    * disableDoubleClickZoom
    * enableCloseButton
    * imageDateControl
    * linksControl
    * panControl
    * pano	
    * position
    * pov
    * scrollwheel	
    * visible	
    * zoomControl		

  * Attr2Options#getControlOptions
    * addressControlOptions
    * panControlOptions
    * zoomControlOptions

  * The way of setting attributes is similiar to `map` directive.

* It must handle all [events] (https://developers.google.com/maps/documentation/javascript/reference#StreetViewPanorama) property. The events and the attribute names are;
    * on-closeclick	
    * on-links_changed 
    * on-pano_changed	
    * on-position_changed	
    * on-pov_changed	
    * on-resize	
    * on-status_changed	
    * on-visible_changed	
    * on-zoom_changed	

* File required to update
   * app/scripts/directives/street-view-panorama.js  
     Please do not forget to document with comments /** .. */
   * testapp/street-view-panorama.html
   * spec/directives/street-view-panorama.js ( If you can)

It is similiar to this pull request. Please take a look at it.

**Comment by rasekar2000** [comment]
Great. this helps .. We will take care of this.

**Comment by allenhwkim** [comment]
any progress?

**Comment by rasekar2000** [comment]
Hi Allen,
  Yes. Stuck in relating Map --> Parent Directive's Controller information into Child Directive . So, move on to other things.. Came back now.. Will keep you posted on the progress in a day or two.
Thanks and Regards,
Raj

**Comment by rasekar2000** [comment]
Hello Allen,

 https://docs.google.com/document/d/1jJOf_jbhrJmXTWHNRRB5FJaNhJq25UmK-Jz4PgLSug0/edit#

 I have documented two approaches in the above document. You will be able to comment on the document itself..

  Option  1) shows the map directive having street-view as an attributes inside map directive.. But it appears the controls are not working as expected in console.log 

```html
<div ng-controller="StreetViewPanoramaCtrl" id="pano" style="width: 70%; height: 100%;float:left">
        <map street-view="StreetViewPanorama(document.querySelector('#pano'), {position:new google.maps.LatLng(40.688738,-74.043871)})" >
                <streetview-panorama  panorama-options="{position: new google.maps.LatLng(40.688649, -74.044689, 40.688606, -74.044513),pov: {heading: 58.387107055378245, pitch: -9.561859294476665},visible: true}" 
on-position_changed="onMapOverlayCompleted()"> </streetview-panorama>                  
          </map>
    </div>
```

  Option 2) shows the streetview-panorama directive having street-view as an attributes inside map directive.. But it shows the blank screen only..

```html
<div ng-controller="StreetViewPanoramaCtrl" id="pano" style="width: 70%; height: 100%;float:left">
         <map zoom="11" center="[40.688738,-74.043871]" >
                <streetview-panorama  panorama-options="{position: new google.maps.LatLng(40.688649, -74.044689, 40.688606, -74.044513),pov: {heading: 270, pitch: 0},visible: true}" on-position_changed="onMapOverlayCompleted()" ></streetview-panorama>                  
          </map>
    </div>
```

Please suggest which option should be pursued. Thanks in advance.

Thanks and Regards,
Raj

**Comment by allenhwkim** [comment]
option 1 is working because street-view is an option of map. This one works without any streetview-directive at the moment.  Why 'streetview-panorama' directive is not working? The directive 'streetview-panorama' is looking for "#pano" which does not exist, thus 'streetview-panorama' must not work

option 2 is the same, inside directive tag, it's looking for "#pano", which we should not.

There are two approaches to show street view.

1. Show streetview in a separate DOM
https://developers.google.com/maps/documentation/javascript/streetview#StreetViewContainers

2. Show streetview within a map
https://developers.google.com/maps/documentation/javascript/streetview#StreetViewOverlays

From your code, I see that directive is written by following  approach #1. However I think you should follow approach #2.

**Comment by rasekar2000** [comment]
Thanks Allen . Will review the same.

**Comment by rasekar2000** [comment]
Hello Allen, 
 I have reviewed your details and [captured it here]

 As i explained earlier, I am finding it difficult to get map object. Its null in the street-view-panorama Directive's isolated scope as well as the mapController object. `link: function(scope, element, attrs, mapController) {`. 
I am not seeing an example that you are using [this option] in any of the program. Yet assuming that the SVP [javascript is ready].  The same is [explained with screenshots].

  Once the map is retrieved and the pano is set then the remaining, properties should go fine as explained here. And the SVP (StreetViewPanorma) methods as well should work fine as they are directly attached to the google maps listener .. 
  Both [SVP properties] and [SVP methods] are captured to the detail in Appendix B

[explained with screenshots]:https://docs.google.com/document/d/1jJOf_jbhrJmXTWHNRRB5FJaNhJq25UmK-Jz4PgLSug0/edit#heading=h.vflfsa4y2o83
[javascript is ready]:http://pastebin.com/fDDGUC2W
[this option]: http://plnkr.co/edit/nuBGNZbKhfDkcZeVXZXN?p=preview
 [SVP properties]: https://docs.google.com/document/d/1jJOf_jbhrJmXTWHNRRB5FJaNhJq25UmK-Jz4PgLSug0/edit#heading=h.y9gr2z4agv1v
 [SVP methods]: https://docs.google.com/document/d/1jJOf_jbhrJmXTWHNRRB5FJaNhJq25UmK-Jz4PgLSug0/edit#heading=h.ro5ehqfdv54a
[captured it here]:https://docs.google.com/document/d/1jJOf_jbhrJmXTWHNRRB5FJaNhJq25UmK-Jz4PgLSug0/edit#heading=h.b6a2mj8krjad

**Comment by allenhwkim** [comment]
ok, i see your issue.  you need to set streetview after a map is initialized. It does not work like others 'obj.setMap'. 

Could you simply send me a pull request if you think its too much? I will work further based on your code.

**Comment by allenhwkim** [comment]
This feature is released with 1.6.0

**Comment by rasekar2000** [comment]
Thanks very much. Let me test them.
